middleware/cluster.go

```diff
@@ -0,0 +1,45 @@
+package middleware
+
+import (
+	"fmt"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/dskit/clusterutil"
+)
+
+// ClusterValidationMiddleware validates that requests have the correct cluster verification label.
+func ClusterValidationMiddleware(cluster string, auxPaths []string, invalidClusters *prometheus.CounterVec, logger log.Logger) Interface {
+	var reB strings.Builder
+	// Allow for a potential path prefix being configured.
+	reB.WriteString(".*/(metrics|debug/pprof.*|ready")
+	for _, p := range auxPaths {
+		reB.WriteString("|" + regexp.QuoteMeta(p))
+	}
+	reB.WriteString(")")
+	reAuxPath := regexp.MustCompile(reB.String())
+
+	return Func(func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			reqCluster := r.Header.Get(clusterutil.ClusterVerificationLabelHeader)
+			if !reAuxPath.MatchString(r.URL.Path) && reqCluster != cluster {
```

Comment:
```suggestion
			if reqCluster != cluster && !reAuxPath.MatchString(r.URL.Path) {
```
Most of the requests are correct, don't waste time doing regexp matching here.